### ISSUE TITLE
Issues after upgrade to engine v1.3.0-15556-473375a

### ISSUE THREAD

**Issue by malikkal** [vmware/vic#7084]
hickeng: The following is the original issue description which is a facet of #7117 and #7091. I have prompted this to an epic for the two distinct upgrade issues that have been identified:
1. containers attached to Open container networks cause various issues after VCH reboot
2. intermittent failure to delete images by ID that were pulled pre-upgrade

Original details:
1. Docker start/stop does not work and results in
`Error response from daemon: all ports are allocated.` 

However, start/stop does work for existing containers that were in 'Created' state.

2. Docker exec does not work and results in
`Error response from daemon: Server error from portlayer: unknown task ID: 168f7b4dd0129d38fec73a032ced341c848eaec9213f8c728f2d3c4373df617c`

3. Cannot delete existing containers, as they would complain as running and stop wouldn’t work either. Force removal would through an error as:
`Error response from daemon: all ports are allocated`.

Interestingly, old cVMs that were up and running as part of docker-compose could be removed using docker-compose down -v. Also, a docker kill stopped an old container, which I could start using docker start later. Subsequent docker stop/start continued **not** to work though..

**Comment by anchal-agrawal** [comment]
Hi @malikkal! For the `Error response from daemon: all ports are allocated.` issue:

* What do this container's port-mappings look like?
* How was this container created/run?

Thanks!

**Comment by malikkal** [comment]
Regardless of the use of -p during run, the error would always appear, when it's on external network. You could notice the difference in output for a ps -a when a container is attached to external network between current and previous versions.

```
PORTS
10.x.y.z:0->0/*
```
This would lead to verbose inspect output #7091. Earlier versions used to be empty.

```
PORTS
10.x.y.z:8080->80/tcp
```
Using NAT, where inspect is okay in new version. Obviously port bindings are different; see below.

```
"PortBindings": {
                "80/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "8080"
                    }
                ]
            }
```
Hope this helps.

**Comment by hickeng** [comment]
I've promoted this to an epic and added related issues as they all derive from the same root cause (assumption here for upgrade, but a reasonable one until verified).

Comment duplicated from https://github.com/vmware/vic/issues/7117#issuecomment-358797679
> @malikkal This is an issue introduced in 1.3.0 that impacts only containers on Open networks and is triggered by restart. One of the docker libraries is taking the port range 0-65535 which are the accessible ports on an open network and decomposing that into a port mapping for each port - this is what you identified in #7091.
When the endpointVM restarts it reapplies NAT forwarding for running containerVMs and mistakenly attempts to apply a NAT rule for each of those 65535 ports, thereby exhausting the port space.

> The application of those NAT rules involves a lot of iteration, many invocations of iptables, and results in a large logging overhead - I believe but have not confirmed that this is what caused your initial upgrade failures; it was actually just taking that long to come up because of the number of port mappings being reconstructed. This portion needs confirming.

> Upgrade hits this specifically because it's synonymous with a restart of a 1.3.0 endpointVM.

**Comment by mhagen-vmware** [comment]
We actually have a fairly extensive test case here:
https://github.com/vmware/vic/blob/master/tests/test-cases/Group11-Upgrade/11-01-Upgrade.robot

It does create a variety of containers and verifies them after upgrade and then rollback too. It is badly dated, which this was supposed to fix a awhile ago (but merges to master have mostly been disabled since I made it):
https://github.com/vmware/vic/pull/6893

@hickeng I am not really seeing the specific scenario that we might be missing that triggers this.  Let's grab some time next week to specifically dig into the various issues that might not be covered by our existing testing together.

**Comment by hickeng** [comment]
@mhagen-vmware We touched on this briefly but for the record here; it's specific to having a running container on a container-network that was configured as Open with --container-network-firewall=net:Open, then restarting the endpointVM.

**Comment by hickeng** [comment]
Closed by #6641
